prescient.el

```diff
@@ -982,6 +982,307 @@ actually just the input, in which case nothing happens."
+     prescient--completion-old-defaults)
+  "Variables used to store old settings.")
+
+;; We keep these "settings" functions separate so that we can use them
+;; in the integration packages without running the entire mode
+;; function.
+
+(defun prescient--completion-apply-completion-settings ()
+  "Apply the recommended settings."
+  (setq completion-styles prescient--completion-recommended-styles
+        completion-category-defaults nil)
+
+  (cl-symbol-macrolet
+      ((category-setting-overrides
+        (alist-get setting
+                   (alist-get category completion-category-overrides))))
+    (cl-loop for (category . overrides)
+             in prescient--completion-recommended-overrides
+             do (cl-loop for (setting . values) in overrides
+                         do (setf category-setting-overrides values)))))
+
+(defun prescient--completion-save-completion-settings ()
+  "Save the old settings."
+  (setq prescient--completion-old-styles completion-styles
+        prescient--completion-old-defaults completion-category-defaults)
+
+  (cl-symbol-macrolet
+      ((category-setting-overrides
+        (alist-get setting
+                   (alist-get category completion-category-overrides))))
+    (setq prescient--completion-old-overrides
+          (cl-loop
+           for (category . overrides)
+           in prescient--completion-recommended-overrides
+           collect
+           `(,category ,@(cl-loop
+                          for (setting . _) in overrides
+                          collect
+                          `(,setting ,@category-setting-overrides)))))))
+
+(defun prescient--completion-restore-completion-settings ()
+  "Restore the old settings."
+  ;; Try to revert back to old settings, or at least not use the
+  ;; `prescient' style.
+  (if (equal completion-styles prescient--completion-recommended-styles)
+      (setq completion-styles prescient--completion-old-styles)
+    (cl-callf2 remq 'prescient completion-styles))
+
+  (cl-loop for (key . val) in prescient--completion-old-defaults
+           unless (alist-get key completion-category-defaults)
+           do (setf (alist-get key completion-category-defaults) val))
+
+  (cl-symbol-macrolet
+      ((category-setting-overrides
+        (alist-get
+         setting
+         (alist-get category completion-category-overrides))))
+    (cl-loop
+     ;; These two trees should have the same structure by this
+     ;; point.
+     for (category . new-overrides)
+     in prescient--completion-recommended-overrides
+     for (_ . old-overrides)
+     in prescient--completion-old-overrides
+     do (cl-loop
+         for (setting . new-values) in new-overrides
+         for (_ . old-values) in old-overrides
+         if (equal category-setting-overrides new-values)
+         do (setf category-setting-overrides old-values)
+         else do (cl-callf2 remq 'prescient category-setting-overrides)))))
+
+;;;###autoload
+(define-minor-mode prescient-completion-mode
```

Comment:
I wonder if we should really expose prescient-completion-mode and prescient-completion-local mode as interactive/public modes for now. I suspect that these modes will lead to confusion since I consider them more as implementation details. I also tend to avoid making APIs public too early.

My suggestion would be to just add two private functions prescient--completion-styles-setup and prescient--completion-styles-revert which could then be used by the public Corfu/Vertico Prescient modes. What do you think?